charms/interface_container_runtime.py

```diff
@@ -0,0 +1,27 @@
+class ContainerRuntimeProvides:
+    """Implements the Provides side of the container-runtime interface."""
+
+    def __init__(self, charm, endpoint):
+        self.charm = charm
+        self.endpoint = endpoint
+
+    @property
+    def relations(self):
+        return self.charm.model.relations[self.endpoint]
+
+    def set_sandbox_image(self, sandbox_image):
+        """Set the sandbox image."""
+        for relation in self.relations:
+            relation.data[self.unit]["sandbox_image"] = sandbox_image
+
+    @property
+    def socket(self):
```

Comment:
I don't know what the downstream users of `socket` are expect.  You COULD parse this with [`urllib.parse.urlparse(...)`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) if you think having it split like that would be useful, but i'm doubting it would

```suggestion
    def socket(self) -> Optional[str]:
```

Reply:
Yeah, we just pass this string as-is to kubelet. I'd rather not parse and reformat it if we don't have to.